SeQuant/domain/eval/eval.hpp

```diff
@@ -97,396 +91,361 @@ void log_eval(Args const&... args) noexcept {
-
-template <typename NodeT, typename Le,
-          std::enable_if_t<IsLeafEvaluator<NodeT, Le>, bool> = true>
-ERPtr evaluate_crust(NodeT const&, Le const&, CacheManager&);
-
-template <typename NodeT, typename Le, typename... Args,
-          std::enable_if_t<IsLeafEvaluator<NodeT, Le>, bool> = true>
-ERPtr evaluate_core(NodeT const& node, Le const& le, Args&&... args) {
   if (node.leaf()) {
-    auto&& [res, time] = timed_eval(le, node);
-    log_eval(node->is_constant()   ? "[CONSTANT] "
-             : node->is_variable() ? "[VARIABLE] "
-                                   : "[TENSOR] ",
-             node->label(), "  ", time.count(), "  ", res->size_in_bytes(),
-             "\n");
-    return res;
+    seconds = timed_eval_inplace([&]() { result = le(node); });
   } else {
-    ERPtr const left =
-        evaluate_crust(node.left(), le, std::forward<Args>(args)...);
-    ERPtr const right =
-        evaluate_crust(node.right(), le, std::forward<Args>(args)...);
-
+    left = evaluate<EvalTrace>(node.left(), le, cache);
+    right = evaluate<EvalTrace>(node.right(), le, cache);
     assert(left);
     assert(right);
 
     std::array<std::any, 3> const ann{node.left()->annot(),
                                       node.right()->annot(), node->annot()};
-
     if (node->op_type() == EvalOp::Sum) {
-      auto&& [res, time] = timed_eval([&]() { return left->sum(*right, ann); });
-      log_eval("[SUM] ", node.left()->label(), " + ", node.right()->label(),
-               " = ", node->label(), "  ", time.count(), "  ",
-               res->size_in_bytes(), "\n");
-      return res;
+      timed_eval_inplace([&]() { result = left->sum(*right, ann); });
     } else {
-      assert(node->op_type() == EvalOp::Prod);
+      assert(node->op_type() == EvalOp::Product);
       auto const de_nest =
           node.left()->tot() && node.right()->tot() && !node->tot();
-
-      auto&& [res, time] = timed_eval([&]() {
-        return left->prod(*right, ann,
-                          de_nest ? TA::DeNest::True : TA::DeNest::False);
+      timed_eval_inplace([&]() {
+        result = left->prod(*right, ann,
+                            de_nest ? TA::DeNest::True : TA::DeNest::False);
       });
+    }
+  }
 
-      log_eval("[PRODUCT] ", node.left()->label(), " * ", node.right()->label(),
-               " = ", node->label(), "  ", time.count(), "  ",
-               res->size_in_bytes(), "\n");
-      return res;
+  assert(result);
+
+  // logging
+  if constexpr (trace(EvalTrace)) {
+    struct {
+      std::string type, annot;
+      size_t bytes;
+    } log;
+    if (node.leaf()) {
+      log.type = node->is_constant()   ? "CONSTANT"
+                 : node->is_variable() ? "VARIABLE"
+                                       : "TENSOR";
+      log.annot = node->label();
+      log.bytes = result->size_in_bytes();
+    } else {
+      log.type = node->is_product() ? "PRODUCT" : node->is_sum() ? "SUM" : "ID";
```

Comment:
if `node.leaf()==false` shouldn't either `node->is_product()` or `node->is_sum()` be true? I.e. shouldn't `"ID"` never be printed?

Reply:
That's true! However, I did not want to put an assertion/throw here because: i) we can simply print the type of evaluation, ii) it is a trace line that could be analyzed further outside the generation.